### ISSUE TITLE
localisation of voxels time series within gray matter masked seed of NiftiSpheresMasker look strange

### ISSUE THREAD

**Issue by salma1601** [nilearn/nilearn#1246]
I extracted voxel time series within seeds using `nifti_spheres_masker._iter_signals_from_spheres`. Since `NiftiSpheresMasker` doesn't have an `inverse_transform` for the moment, I tried to recover my image using directly `NiftiMasker` and identifiying which voxels actually have the same signals as my within seeds voxels.

Things look OK when I don't use a `mask_img` in my `NiftiSpheresMasker`, but when I use a grey matter mask to intersect my seed with grey matter, I have very strange results, as if voxels were sliced with a step of 3. I find this very strange, and actually don't know if it's my code or a `NiftiSpheresMasker` bug, so posting it just to report. I tried this with different seed/radius, and it's the same strory.
![rtj_sphere_no_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17700526/63ec96d4-63c6-11e6-8842-ad4c5c82e85f.png)
![rtj_sphere_with_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17700530/676045f4-63c6-11e6-9b4d-0b1bc7895b03.png)

I am providing my code if needed for checking, it's very long and ugly :$

``` python
from nilearn import datasets, input_data
adhd_dataset = datasets.fetch_adhd(n_subjects=1)
func_filename = adhd_dataset.func[0]
rtj_coords = (-46, -68, 32)
radius = 8.
brain_masker = input_data.NiftiMasker(standardize=True)
brain_voxels_time_series = brain_masker.fit_transform(func_filename)

# Manual extraction of voxel timeseries within the masked sphere
import nibabel
import numpy as np
from nilearn.input_data.nifti_spheres_masker import _iter_signals_from_spheres
sphere_masker = input_data.NiftiSpheresMasker(
    [rtj_coords], standardize=True, radius=radius)
sphere_masker.fit()
func_img = nibabel.load(func_filename)
sphere_time_series = _iter_signals_from_spheres(
    sphere_masker.seeds_, func_img, sphere_masker.radius,
    sphere_masker.allow_overlap, mask_img=sphere_masker.mask_img).next()
print('Check no zero time series: min std is {0}'.format(
    sphere_time_series.std(axis=0).min()))

# Form sphere image
sphere_indicator = np.zeros(brain_voxels_time_series.shape[1:])
indices = []
for ts in (sphere_time_series / sphere_time_series.std(axis=0)).T:
    voxel_index = np.argmin(
        np.linalg.norm(brain_voxels_time_series - ts[:, np.newaxis], axis=0))
    indices.append(voxel_index)

sphere_indicator[indices] = 1.
assert(np.sum(sphere_indicator) == sphere_time_series.shape[1])
sphere_indicator_img = brain_masker.inverse_transform(sphere_indicator)

# Repeat wile masking with GM 
icbm152_grey_mask = datasets.fetch_icbm152_brain_gm_mask()
gm_masker = input_data.NiftiMasker(
    mask_img=icbm152_grey_mask, standardize=True)
gm_voxels_time_series = gm_masker.fit_transform(func_filename)
gm_seed_masker = input_data.NiftiSpheresMasker(
    [rtj_coords], standardize=True, radius=radius, mask_img=icbm152_grey_mask)
gm_seed_masker.fit()
gm_sphere_time_series = _iter_signals_from_spheres(
    gm_seed_masker.seeds_, func_img, gm_seed_masker.radius,
    gm_seed_masker.allow_overlap, mask_img=gm_seed_masker.mask_img).next()
gm_sphere_indicator = np.zeros(gm_voxels_time_series.shape[1:])
indices = []
for ts in (gm_sphere_time_series / gm_sphere_time_series.std(axis=0)).T:
    voxel_index = np.argmin(
        np.linalg.norm(gm_voxels_time_series - ts[:, np.newaxis], axis=0))
    indices.append(voxel_index)

gm_sphere_indicator[indices] = 1.
assert(np.sum(gm_sphere_indicator) == gm_sphere_time_series.shape[1])
gm_sphere_indicator_img = gm_masker.inverse_transform(gm_sphere_indicator)

from nilearn import _utils, plotting
for img, title in zip([sphere_indicator_img, gm_sphere_indicator_img],
                      ['unmasked', 'GM masked']):
    display = plotting.plot_roi(_utils.check_niimg(img, atleast_4d=True),
                                display_mode='z',
                                cut_coords=range(24, 45, 3),
                                title=title, bg_img=icbm152_grey_mask)
plotting.show()
```

**Comment by AlexandreAbraham** [comment]
I don't have much time to debug this but I'll give it a quick try. Thanks for the script. At first glance, this looks like a resampling issue.

**Comment by AlexandreAbraham** [comment]
I have a MemoryError with your script... The magic should happend at the beginning of `_apply_mask_and_get_affinity` function. Could you take a look at mask_coords and see if there is something fishy?

**Comment by salma1601** [comment]
> MemoryError with your script

computing all voxels timeseries is memory consuming I guess ? I actually don't have this pb
>Could you take a look at mask_coords and see if there is something fishy

Ok

**Comment by salma1601** [comment]
`mask_coords` seem to be sliced by a stem of 3, for instance here are the first 10 (I changed the radius to 3 to accelerate)
>[[90, -126, -72], [90, -126, -69], [90, -126, -66], [90, -126, -63], [90, -126, -60], [90, -126, -57], [90, -126, -54], [90, -126, -51], [90, -126, -48], [90, -126, -45]]

**Comment by salma1601** [comment]
do you want me to try fix this issue or you want to fix it ?

**Comment by AlexandreAbraham** [comment]
Those are coordinates of the center of the voxels. The distance of 3 between them is expected since it should be equal to your resolution. However, that makes a lot of points for a radius of 3mm...

**Comment by salma1601** [comment]
`mask_coords` has shape `(271633, 3)`

**Comment by salma1601** [comment]
The function `_apply_mask_and_get_affinity` seems OK for me: I checked my signal values for radius `3mm` and they are quite the same with and without masking! However the problem persists in my plotting. Could it be related to the `NiftiMasker` `inverse_transform` or the plotting itself ??
![rtj_sphere_3mm_no_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17707929/ef1c8c7a-63e2-11e6-94b4-ff02bad4af62.png)![rtj_sphere_3mm_with_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17707933/f3261cd2-63e2-11e6-8b13-1daf595514a8.png)

**Comment by salma1601** [comment]
I think `_apply_mask_and_get_affinity` is OK: I have exactly the same time points for a radius of 3mm with and without GM mask !
But still the same issue when plotting.
![rtj_sphere_3mm_no_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17709519/ba90cc94-63e9-11e6-9a35-e98601a3d520.png)![rtj_sphere_3mm_with_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17709526/c34a23f8-63e9-11e6-86ab-077a3b8785d3.png)

 Can the pb be related to the `NiftiMasker` `inverse_transform` ??

**Comment by AlexandreAbraham** [comment]
Could you save the img that you are plotting? I was able to reproduce a plot that looks like yours by messing with data types, when int data is used and there is interpolation, but your data seems to be float.

**Comment by salma1601** [comment]
>but your data seems to be float.

my grey matter mask and my seed mask are int. But I converted everything to float and it does change nothing.

>Could you save the img that you are plotting? 

What do you mean ? I saved in .png and in .pdf but still the same problem

**Comment by AlexandreAbraham** [comment]
I mean this:
```python
sphere_indicator_img.to_filename('good.nii.gz')
gm_sphere_indicator_img.to_filename('not_good.nii.gz')
```

**Comment by salma1601** [comment]
with fsl looks the same as the plotted image with nilearn plotting

**Comment by AlexandreAbraham** [comment]
OK, I have the answer. So, what happens is that you have problems with your affines. ICBM template resolution is 1mm. When you create your masker with it, all the data processed through it will be resampled to 1mm. That means that the signals in your functional image is interpolated.

When you use the iterator on spheres, you iterate on the 3mm image which means that you will never grab the signal of the voxels in 1mm resolution.

I don't know if it is clear but here are a few ways to solve the problem:
```python
gm_masker = input_data.NiftiMasker(
    mask_img=icbm152_grey_mask, standardize=True, target_affine=[3, 3, 3])
```
or
```python
resampled_img = gm_masker.inverse_transform(gm_masker.transform(func_img))

gm_sphere_time_series = _iter_signals_from_spheres(
    gm_seed_masker.seeds_, resampled_img, gm_seed_masker.radius,
    gm_seed_masker.allow_overlap, mask_img=gm_seed_masker.mask_img).next()
```

Good practice: **always** resample all your stuff to the same affine before processing.

**Comment by salma1601** [comment]
Ok, thank you very much !

**Comment by salma1601** [comment]
The first code
```python
gm_masker = input_data.NiftiMasker(
    mask_img=icbm152_grey_mask, standardize=True, target_affine=3*np.eye(3))
```
 throws an error
>Given mask is not made of 2 values

may be related to #1171 ?
But the second one works fine !
![rtj_sphere_6mm_no_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17755381/0d8adf66-64da-11e6-926a-fcbf31cf9cd2.png)![rtj_sphere_6mm_with_gm_mask](https://cloud.githubusercontent.com/assets/7080143/17755398/19ab6464-64da-11e6-922d-516b022911e7.png)

**Comment by bthirion** [comment]
Can't reproduce any more closing.
